content/blog/2020-12-18-brilc.md

```diff
@@ -0,0 +1,248 @@
+  whatever decisions we need to make must be evaluated by code that was emitted
+  statically.
+
+### Contributions
+
+To explore these issues, I built a new compiler from Bril to LLVM that supports the
+[memory extension][memex] (due to Zagieboylo, Doenges).
+
+Before getting into any garbage collection, I needed a little more expressive
+language for heap manipulation than just pointers to primitive arrays. For this,
+I built [a compiler][brilc] that implements a new *struct extention* for Bril, which allows
+the use of structs similar to those in C.
+
+
+# A Struct Extension for Bril
+
+Programs using this extension have a second top-level key (i.e., in addition to
+`"functions"`), called `"structs"`, which define C-like struct datatypes for use
+in the program. A struct definition has only two keys:
+* `name`: a name for the struct type.
+* `mbrs`: a list of dictionaries that represent the members of the struct type.
+          Each dictionary has two keys, `type`, which gives a valid bril type,
+          and `name`. (Observe that these are the same two fields that an
+          argument to a function has).
+
+## Warm-up Example
+To use the standard first struct example, we might define a struct to represent
+a point in `$\textbf{Z}^2$`. (Note that we narrowly miss one of the three
+disallowed names for a struct type, `int`, `bool`, and `ptr`).
+```
+struct point = {
+    x: int;
+    y: int;
+}
+```
+which would have the JSON representation:
+```
+{"name": "point", "mbrs":[{"type": "int", "name": "x"}, {"type": "int", "name": "y"}]}
+```
+For the rest of this post, we'll only use the text representation for
+definitions.
+
+## Memory Allocation for Structs
+Instances of structs are only allowed to be allocated on the heap, using the 
+memory extension .  Thus, we can use
+`alloc` without modification to get a space for an instance of a struct on
+the heap, or to get really exciting, a whole array. Importantly, the struct is
+not automatically initialized. It is illegal to `load` from any struct member 
+that has not been previously written to.
+
+So we can allocate a `point` with:
+```
+one: int = const 1
+ptr<point> p = alloc one
+```
+
+To access the members of the struct, we only need a way to get a name for the
+members, since we can read and write using `load` and `store` from the normal
+memory extension. For this, we a new value instructions:
+
+* `getmbr`, which takes two arguments: a pointer to a struct, and the name of a
+  member of the struct. It returns a pointer to the requested element, with
+  appropriate pointer type. Since we can only ever have pointers to structs,
+  this is always possible.
+
+In text form, this would look like this:
+```
+<dest>: <ty> = getmbr <ptr> <mbr>
+```
+If this evokes LLVM's `getelementptr` in your mind, you've got it! (That is how
+this instruction is implemented). In the future, it might be convenient to
+provide a "dot-notation" as syntactic sugar for member access.
```

Comment:
It might, but for an IR, I think the GEP-like semantics are good… this separates the struct-specific behavior (interacting with members) from the general concept of loading and storing.